File: samples/snippets/cpp/VS_Snippets_CLR_System/system.Security.Cryptography.KeySizes/CPP/members.cpp
// This sample demonstrates how to use each member of the KeySizes class.
//<Snippet1>

using namespace System;
using namespace System::Security::Cryptography;

namespace CryptographySample
{
    ref class KeySizesMembers
    {
    public:
        static void Work()
        {
            // Initializes a new instance of the KeySizes class 
            // with the specified key values.
            //<Snippet2>
            int minSize = 64;
            int maxSize = 1024;
            int skipSize = 64;
            KeySizes^ keySizes = 
                gcnew KeySizes(minSize, maxSize, skipSize);
            //</Snippet2>

            // Show the values of the keys.
            ShowKeys(gcnew array<KeySizes^>(1) {keySizes},
                "Custom Keys");
                 
            // Create a new symmetric algorithm and display its 
            // key values.
            Aes^ aes = Aes::Create();
            ShowKeys(aes->LegalKeySizes, aes->ToString());
            Console::WriteLine("aes.blocksize:{0}", 
                    aes->BlockSize);

            // Create a new RSA algorithm and display its key values.
            RSA^ rsa = RSA::Create();
            ShowKeys(rsa->LegalKeySizes, rsa->ToString());
            Console::WriteLine("RSA KeySize ="
                " {0}",
                    rsa->KeySize);

            Console::WriteLine("This sample completed successfully; " 
                   "press Enter to exit.");
            Console::ReadLine();
        }

    private:
        // Display specified KeySize properties to the console.
        static void ShowKeys(array <KeySizes^>^ keySizes,
            String^ objectName)
        {
            // Retrieve the first KeySizes in the array.
            KeySizes^ firstKeySize = keySizes[0];

            // Retrieve the minimum key size in bits.
            //<Snippet3>
            int minKeySize = firstKeySize->MinSize;
            //</Snippet3>

            // Retrieve the maximum key size in bits.
            //<Snippet4>
            int maxKeySize = firstKeySize->MaxSize;
            //</Snippet4>

            // Retrieve the interval between valid key size in bits.
            //<Snippet5>
            int skipKeySize = firstKeySize->SkipSize;
            //</Snippet5>

            Console::Write("\n KeySizes retrieved from the ");
            Console::WriteLine("{0} object.", objectName);
            Console::WriteLine("Minimum key size bits: {0}", 
                minKeySize);
            Console::WriteLine("Maximum key size bits: {0}", 
                maxKeySize);
            Console::WriteLine("Interval between key size bits: {0}", 
                skipKeySize);            
        }
    };
}

using namespace CryptographySample;

int main()
{
    KeySizesMembers::Work();
}
//
// This sample produces the following output:
//
// KeySizes retrieved from the Custom Keys object.
// Minimum key size bits: 64
// Maximum key size bits: 1024
// Interval between key size bits: 64
// 
// KeySizes retrieved from the 
//  System.Security.Cryptography.Aes object.
// Minimum key size bits: 128
// Maximum key size bits: 256
// Interval between key size bits: 64
// aes.blocksize:128
// 
// KeySizes retrieved from the
// System.Security.Cryptography.RSA object.
// Minimum key size bits: 512
// Maximum key size bits: 16384
// Interval between key size bits: 64
// RSA KeySize = 2048
// This sample completed successfully; press Enter to exit.
//</Snippet1>

File: snippets/csharp/System.Security.Policy/PolicyStatement/Overview/members.cs
// This sample demonstrates how to use each member of the PolicyStatement
// class.
//<Snippet1>
using System;
using System.Security;
using System.Security.Policy;
using System.Security.Principal;
using System.Security.Permissions;

class Members
{
    [STAThread]
    static void Main(string[] args)
    {
        // Create two new policy statements.
        PolicyStatement policyStatement = firstConstructorTest();
        PolicyStatement policyStatement2 = secondConstructorTest();

        // Add attributes to the first policy statement.
        //<Snippet4>
        policyStatement.Attributes = PolicyStatementAttribute.All;
        //</Snippet4>

        // Create a copy of the first policy statement.
        PolicyStatement policyStatementCopy = createCopy(policyStatement);
        addXmlMember(ref policyStatementCopy);

        summarizePolicyStatment(policyStatement);
        Console.WriteLine("This sample completed successfully; " +
            "press Enter to exit.");
        Console.ReadLine();
    }

    // If a class attribute is not found in the specified PolicyStatement,
    // add a child XML element with an added class attribute.
    private static void addXmlMember(ref PolicyStatement policyStatement)
    {
        //<Snippet6>
        SecurityElement xmlElement = policyStatement.ToXml();
        //</Snippet6>
        if (xmlElement.Attribute("class") == null)
        {
            //<Snippet7>
            SecurityElement newElement =
                new SecurityElement("PolicyStatement");
            newElement.AddAttribute("class", policyStatement.ToString());
            newElement.AddAttribute("version","1.1");

            newElement.AddChild(new SecurityElement("PermissionSet"));

            policyStatement.FromXml(newElement);
            //</Snippet7>

            Console.Write("Added the class attribute and modified its ");
            Console.WriteLine("version number.\n" + newElement.ToString());
        }
    }

    // Verify that the type of the specified object is a PolicyStatement type
    // then create a copy of the object.
    private static PolicyStatement createCopy(Object sourceObject)
    {
        PolicyStatement returnedStatement = new PolicyStatement(null);
        // Compare specified object type with the PolicyStatement type.
        //<Snippet8>
        if (sourceObject.GetType().Equals(typeof(PolicyStatement)))
            //</Snippet8>
        {
            returnedStatement = getCopy((PolicyStatement)sourceObject);
        }
        else
        {
            throw new ArgumentException("Expected the PolicyStatement type.");
        }

        return returnedStatement;
    }

    // Return a copy of the specified PolicyStatement if the result of the
    // Copy command is an equivalent object. Otherwise, return the
    // original PolicyStatement object.
    private static PolicyStatement getCopy(PolicyStatement policyStatement)
    {
        // Create an equivalent copy of the policy statement.
        //<Snippet9>
        PolicyStatement policyStatementCopy = policyStatement.Copy();
        //</Snippet9>

        // Compare the specified objects for equality.
        //<Snippet10>
        if (!policyStatementCopy.Equals(policyStatement))
            //</Snippet10>
        {
            return policyStatementCopy;
        }
        else
        {
            return policyStatement;
        }
    }

    // Summarize the attributes of the specified PolicyStatement on the
    // console window.
    private static void summarizePolicyStatment(
        PolicyStatement policyStatement)
    {
        // Retrieve the class path for policyStatement.
        //<Snippet11>
        string policyStatementClass = policyStatement.ToString();
        //</Snippet11>

        //<Snippet12>
        int hashCode = policyStatement.GetHashCode();
        //</Snippet12>

        string attributeString = "";
        // Retrieve the string representation of the PolicyStatement
        // attributes.
        //<Snippet13>
        if (policyStatement.AttributeString != null)
        {
            attributeString = policyStatement.AttributeString;
        }
        //</Snippet13>

        // Write a summary to the console window.
        Console.WriteLine("\n*** " + policyStatementClass + " summary ***");
        Console.Write("This PolicyStatement has been created with hash ");
        Console.Write("code(" + hashCode + ") ");

        Console.Write("and contains the following attributes: ");
        Console.WriteLine(attributeString);
    }
}
//
// This sample produces the following output:
//
// Added the class attribute and modified the version number.
// <PolicyStatement class="System.Security.Policy.PolicyStatement"
//                  version="1.1">
//    <PermissionSet/>
// </PolicyStatement>
//
// *** System.Security.Policy.PolicyStatement summary ***
// PolicyStatement has been created with hash code(20) containing the
// following attributes: Exclusive LevelFinal
// This sample completed successfully; press Enter to exit.
//</Snippet1>
